sklearn/metrics/tests/test_common.py

```diff
@@ -1022,54 +1053,64 @@ def check_sample_weight_invariance(name, metric, y1, y2):
                                                   sample_weight]))
 
 
-def test_sample_weight_invariance(n_samples=50):
+@pytest.mark.parametrize(
+        'name',
+        (set(ALL_METRICS).intersection(REGRESSION_METRICS)
+         - set(METRICS_WITHOUT_SAMPLE_WEIGHT)))
+def test_regression_sample_weight_invariance(name):
+    n_samples = 50
     random_state = check_random_state(0)
     # regression
     y_true = random_state.random_sample(size=(n_samples,))
     y_pred = random_state.random_sample(size=(n_samples,))
-    for name in ALL_METRICS:
-        if name not in REGRESSION_METRICS:
-            continue
-        if name in METRICS_WITHOUT_SAMPLE_WEIGHT:
-            continue
-        metric = ALL_METRICS[name]
-        yield check_sample_weight_invariance, name, metric, y_true, y_pred
+    metric = ALL_METRICS[name]
+    check_sample_weight_invariance(name, metric, y_true, y_pred)
 
+
+@pytest.mark.parametrize(
+        'name',
+        (set(ALL_METRICS) - set(REGRESSION_METRICS)
+         - set(METRICS_WITHOUT_SAMPLE_WEIGHT) - set(METRIC_UNDEFINED_BINARY)))
+def test_binary_sample_weight_invariance(name):
     # binary
+    n_samples = 50
     random_state = check_random_state(0)
     y_true = random_state.randint(0, 2, size=(n_samples, ))
     y_pred = random_state.randint(0, 2, size=(n_samples, ))
     y_score = random_state.random_sample(size=(n_samples,))
-    for name in ALL_METRICS:
-        if name in REGRESSION_METRICS:
-            continue
-        if (name in METRICS_WITHOUT_SAMPLE_WEIGHT or
-                name in METRIC_UNDEFINED_BINARY):
-            continue
-        metric = ALL_METRICS[name]
-        if name in THRESHOLDED_METRICS:
-            yield check_sample_weight_invariance, name, metric, y_true, y_score
-        else:
-            yield check_sample_weight_invariance, name, metric, y_true, y_pred
+    metric = ALL_METRICS[name]
+    if name in THRESHOLDED_METRICS:
+        check_sample_weight_invariance(name, metric, y_true, y_score)
+    else:
+        check_sample_weight_invariance(name, metric, y_true, y_pred)
+
 
+@pytest.mark.parametrize(
+        'name',
+        (set(ALL_METRICS) - set(REGRESSION_METRICS)
+         - set(METRICS_WITHOUT_SAMPLE_WEIGHT)
+         - METRIC_UNDEFINED_BINARY_MULTICLASS))
```

Comment:
Adding set to keep consistent? I'm fine with current version though.

Reply:
Yes, this is a bit awkward because `ALL_METRICS` and `REGRESSION_METRICS` are dicts, while `METRICS_WITHOUT_SAMPLE_WEIGHT` and all the rest are lists, and finally `METRIC_UNDEFINED_BINARY_MULTICLASS` is a set.

I changed the ones that were lists to sets which removes the need to cast them to set in parametrizations (since we are doing a number of set operations here).